src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -209,16 +213,36 @@ private ImmutableHashSet<ProjectId> GetProjectsThatDirectlyDependOnThisProject_N
 
     private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeReverseReferencesMap()
     {
-        var reverseReferencesMap = new Dictionary<ProjectId, HashSet<ProjectId>>();
+        using var _ = PooledDictionary<ProjectId, ImmutableHashSet<ProjectId>.Builder>.GetInstance(out var reverseReferencesMap);
 
         foreach (var (projectId, references) in _referencesMap)
         {
             foreach (var referencedId in references)
-                reverseReferencesMap.MultiAdd(referencedId, projectId);
+            {
+                if (!reverseReferencesMap.TryGetValue(referencedId, out var builder))
+                {
+                    builder = s_reverseReferencesBuilderPool.Allocate();
+                    reverseReferencesMap.Add(referencedId, builder);
+                }
+
+                builder.Add(projectId);
+            }
         }
 
         return reverseReferencesMap
```

Comment:
consider just writing the loop by hand, instead of select methods.

Reply:
yes.  please do that.  the mix of builders and linq is unplsaeant :)

Reply:
This look better? Will start off another perf run if so to make sure I get accurate numbers.